html/qhull.man
.\"  This is the Unix manual page for qhull, written in nroff, the standard
.\"  manual formatter for Unix systems.  To format it, type
.\"
.\"  nroff -man qhull.man
.\"
.\"  This will print a formatted copy to standard output.  If you want
.\"  to ensure that the output is plain ASCII, free of any control
.\"  characters that nroff uses for underlining etc, pipe the output
.\"  through "col -b":
.\"
.\"  nroff -man qhull.man | col -b
.\"
.\"  Warning: a leading quote "'" or dot "." will not format correctly
.\"
.TH qhull 1 "2003/12/30" "Geometry Center"
.SH NAME
qhull \- convex hull, Delaunay triangulation, Voronoi diagram,
halfspace intersection about a point, hull volume, facet area
.SH SYNOPSIS
.nf
qhull- compute convex hulls and related structures
    input (stdin): dimension, #points, point coordinates
    first comment (non-numeric) is listed in the summary
    halfspace: use dim plus one with offsets after coefficients

options (qh-quick.htm):
    d    - Delaunay triangulation by lifting points to a paraboloid
    d Qu - furthest-site Delaunay triangulation (upper convex hull)
    v    - Voronoi diagram as the dual of the Delaunay triangulation
    v Qu - furthest-site Voronoi diagram
    H1,1 - Halfspace intersection about [1,1,0,...] via polar duality
    Qt   - triangulated output
    QJ   - joggled input instead of merged facets
    Tv   - verify result: structure, convexity, and point inclusion
    .    - concise list of all options
    -    - one-line description of each option
    -?   - this message
    -V   - version

Output options (subset):
    s    - summary of results (default)
    i    - vertices incident to each facet
    n    - normals with offsets
    p    - vertex coordinates (if 'Qc', includes coplanar points)
           if 'v', Voronoi vertices
    FA   - report total area and volume
    Fp   - halfspace intersections
    FS   - total area and volume
    Fx   - extreme points (convex hull vertices)
    G    - Geomview output (2-d, 3-d and 4-d)
    m    - Mathematica output (2-d and 3-d)
    o    - OFF format (if 'v', outputs Voronoi regions)
    QVn  - print facets that include point n, -n if not
    TI file - input file, may be enclosed in single quotes
    TO file - output file, may be enclosed in single quotes

examples:
    rbox D4 | qhull Tv                        rbox 1000 s | qhull Tv s FA
    rbox 10 D2 | qhull d QJ s i TO result     rbox 10 D2 | qhull v Qbb Qt p
    rbox 10 D2 | qhull d Qu QJ m              rbox 10 D2 | qhull v Qu QJ o
    rbox c d D2 | qhull Qc s f Fx | more      rbox c | qhull FV n | qhull H Fp
    rbox d D12 | qhull QR0 FA                 rbox c D7 | qhull FA TF1000
    rbox y 1000 W0 | qhull Qc                 rbox c | qhull n
.fi

 - html manual:    html/index.htm
 - installation:   README.txt
 - see also:       COPYING.txt, REGISTER.txt, Changes.txt
 - WWW:            <http://www.qhull.org>
 - GIT:            <git@github.com:qhull/qhull.git>
 - news:           <http://www.qhull.org/news>
 - Geomview:       <http://www.geomview.org>
 - news group:     <news:comp.graphics.algorithms>
 - FAQ:            <http://www.faqs.org/faqs/graphics/algorithms-faq/>
 - email:          qhull@qhull.org
 - bug reports:    qhull_bug@qhull.org

The sections are:
 - INTRODUCTION
 - DESCRIPTION, a description of Qhull
 - IMPRECISION, how Qhull handles imprecision
 - OPTIONS
 -    Input and output options
 -    Additional input/output formats
 -    Precision options
 -    Geomview options
 -    Print options
 -    Qhull options
 -    Trace options
 - BUGS
 - E-MAIL
 - SEE ALSO
 - AUTHORS
 - ACKNOWLEGEMENTS

This man page briefly describes all Qhull options.  Please report
any mismatches with Qhull's html manual (html/index.htm).

.PP
.SH INTRODUCTION
Qhull is a general dimension code for computing convex hulls, Delaunay
triangulations, Voronoi diagram, furthest\[hy]site Voronoi diagram,
furthest\[hy]site Delaunay triangulations, and
halfspace intersections about a point.  It implements the Quickhull algorithm for
computing the convex hull.  Qhull handles round\[hy]off errors from floating
point arithmetic.  It can approximate a convex hull.

The program includes options for hull volume, facet area, partial hulls,
input transformations, randomization, tracing, multiple output formats, and
execution statistics.  The program can be called from within your application.
You can view the results in 2\[hy]d, 3\[hy]d and 4\[hy]d with Geomview.
.PP
.SH DESCRIPTION
.PP
The format of input is the following: first line contains the dimension,
second line contains the number of input points, and point coordinates follow.
The dimension and number of points can be reversed.
Comments and line breaks are ignored.  A comment starts with a
non\[hy]numeric character and continues to the end of line.  The first comment
is reported in summaries and statistics.
Error reporting is
better if there is one point per line.
.PP
The default printout option is a short summary. There are many
other output formats.
.PP
Qhull implements the Quickhull algorithm for convex hull. This algorithm combines
the 2\[hy]d Quickhull algorithm with the n\[hy]d beneath\[hy]beyond algorithm
[c.f., Preparata & Shamos '85].
It is similar to the randomized algorithms of Clarkson and
others [Clarkson et al. '93].  The main
advantages of Quickhull are output sensitive performance, reduced
space requirements, and automatic handling of precision problems.
.PP
The data structure produced by Qhull consists of vertices, ridges, and facets.
A vertex is a point of the input set.  A ridge is a set of d vertices
and two neighboring facets.  For example in 3\[hy]d, a ridge is an edge of the
polyhedron.  A facet is a set of ridges, a set of neighboring facets, a set
of incident vertices, and a hyperplane equation.  For simplicial facets, the
ridges are defined by the vertices and neighboring facets.  When Qhull
merges two facets, it produces a non\[hy]simplicial
facet.  A non\[hy]simplicial facet has more than d neighbors and may share more than
one ridge with a neighbor.
.PP
.SH IMPRECISION
.PP
Since Qhull uses floating point arithmetic, roundoff error may occur for each
calculation.  This causes  problems
for most geometric algorithms.
.PP
Qhull automatically sets option 'C\-0' in 2\[hy]d, 3\[hy]d, and 4\[hy]d, or
option 'Qx' in 5\[hy]d and higher.  These options handle precision problems
by merging facets.  Alternatively, use option 'QJ' to joggle the
input.
.PP
With 'C\-0', Qhull merges non\[hy]convex
facets while constructing the hull. The remaining facets are
clearly convex. With 'Qx', Qhull merges
coplanar horizon facets, flipped facets, concave facets and
duplicated ridges.  It merges coplanar facets after constructing
the hull.
With 'Qx', coplanar points may be missed, but it
appears to be unlikely.
.PP
To guarantee triangular output, joggle the input with option 'QJ'.  Facet
merging will not occur.
.SH OPTIONS
.PP
To get a list of the most important options, execute 'qhull -?'.
To get a complete list of options, execute 'qhull \-'.
To get a complete, concise list of options, execute 'qhull .'.

Options can be in any order.
Capitalized options take an argument (except 'PG' and 'F' options).
Single letters are used for output formats and precision constants.  The
other options are grouped into menus: output formats ('F'),
Geomview output ('G'),
printing ('P'), Qhull control ('Q'), and tracing ('T').
.TP
Main options:
.TP
default
Compute the convex hull of the input points.  Report a summary of
the result.
.TP
d
Compute the Delaunay triangulation by lifting the input points to a
paraboloid.  The 'o' option prints the input points and facets.
The 'QJ' option guarantees triangular output.  The 'Ft'
option prints a triangulation.  It adds points (the centrums) to non\[hy]simplicial
facets.
.TP
v
Compute the Voronoi diagram from the Delaunay triangulation.
The 'p' option prints the Voronoi vertices.
The 'o' option prints the Voronoi vertices and the
vertices in each Voronoi region.  It lists regions in
site ID order.
The 'Fv' option prints each ridge of the Voronoi diagram.
The first or zero'th vertex
indicates the infinity vertex.  Its coordinates are
qh_INFINITE (\-10.101).  It indicates unbounded Voronoi
regions or degenerate Delaunay triangles.
.TP
Hn,n,...
Compute halfspace intersection about [n,n,0,...].
The input is a set of halfspaces
defined in the same format as 'n', 'Fo', and 'Fi'.
Use 'Fp' to print the intersection points.  Use 'Fv'
to list the intersection points for each halfspace.  The
other output formats display the dual convex hull.

The point [n,n,n,...] is a feasible point for the halfspaces, i.e.,
a point that is inside all
of the halfspaces (Hx+b <= 0).  The default coordinate value is 0.

The input may start with a feasible point.  If so, use 'H' by itself.
The input starts with a feasible point when the first number is the dimension,
the second number is "1", and the coordinates complete a line.  The 'FV'
option produces a feasible point for a convex hull.
.TP
d Qu
Compute the furthest\[hy]site Delaunay triangulation from the upper
convex hull.  The 'o' option prints the input points and facets.
The 'QJ' option guarantees triangular output.  You can also use 'Ft'
to triangulate via the centrums of non\[hy]simplicial
facets.
.TP
v Qu
Compute the furthest\[hy]site Voronoi diagram.
The 'p' option prints the Voronoi vertices.
The 'o' option prints the Voronoi vertices and the
vertices in each Voronoi region.
The 'Fv' option prints each ridge of the Voronoi diagram.
The first or zero'th vertex
indicates the infinity vertex at infinity.  Its coordinates are
qh_INFINITE (\-10.101).  It indicates unbounded Voronoi regions
and degenerate Delaunay triangles.
.PP
.TP
Input/Output options:
.TP
f
Print all facets and all fields of each facet.
.TP
G
Output the hull in Geomview format.  For imprecise hulls,
Geomview displays the inner and outer hull.  Geomview can also
display points, ridges, vertices, coplanar points, and
facet intersections.  See below for a list of options.

For Delaunay triangulations, 'G' displays the
corresponding paraboloid.  For halfspace intersection, 'G' displays the
dual polytope.
.TP
i
Output the incident vertices for each facet.
Qhull prints the number of facets followed by the
vertices of each facet.  One facet is printed per line.  The numbers
are the 0\[hy]relative indices of the corresponding input points.
The facets
are oriented.

In 4d and higher,
Qhull triangulates non\[hy]simplicial facets.  Each apex (the first vertex) is
a created point that corresponds to the facet's centrum.  Its index is greater
than the indices of the input points.  Each base
corresponds to a simplicial ridge between two facets.
To print the vertices without triangulation, use option 'Fv'.
To print the centrum coordinates, use option 'Ft'.
The centrum indices 
for option 'i' are one more than the centrum indices for option 'Ft'.
.TP
m
Output the hull in Mathematica format.  Qhull writes a Mathematica file for 2\[hy]d and 3\[hy]d
convex hulls and for 2\[hy]d Delaunay triangulations.   Qhull produces a list of objects
that you can assign to a variable in Mathematica, for example:
"list= << <outputfilename> ". If the object is 2\[hy]d, it can be
visualized by "Show[Graphics[list]] ". For 3\[hy]d objects the command is
"Show[Graphics3D[list]]".
.TP
n
Output the normal equation for each facet.
Qhull prints the dimension (plus one), the number of facets,
and the normals for each facet.  The facet's offset follows its
normal coefficients.
.TP
o
Output the facets in OFF file format.
Qhull prints the dimension, number of points, number
of facets, and number of ridges.  Then it prints the coordinates of
the input points and the vertices for each facet.  Each facet is on
a separate line.  The first number is the number of vertices.  The
remainder are the indices of the corresponding points.  The vertices are
oriented in 2\[hy]d, 3\[hy]d, and in simplicial facets.

For 2\[hy]d Voronoi diagrams,
the vertices are sorted by adjacency, but not oriented.  In 3\[hy]d and higher,
the Voronoi vertices are sorted by index.
See the 'v' option for more information.
.TP
p
Output the coordinates of each vertex point.
Qhull prints the dimension, the number of points,
and the coordinates for each vertex.
With the 'Gc' and 'Gi' options, it also prints coplanar
and interior points.  For Voronoi diagrams, it prints the coordinates
of each Voronoi vertex.
.TP
s
Print a summary to stderr.  If no output options
are specified, a summary goes to stdout.  The summary lists
the number of input points, the dimension, the number of vertices
in the convex hull, the number of facets in the convex hull, the
number of good facets (if 'Pg'), and statistics.

The last two statistics (if needed) measure the maximum distance
from a point or vertex to a
facet.  The number in parenthesis (e.g., 2.1x) is the ratio between the
maximum distance and the worst\[hy]case distance due to merging
two simplicial facets.
.PP
.TP
Precision options
.TP
An
Maximum angle given as a cosine.  If the angle between a pair of facet
normals
is greater than n, Qhull merges one of the facets into a neighbor.
If 'n' is negative, Qhull tests angles after adding
each point to the hull (pre\[hy]merging).
If 'n' is positive, Qhull tests angles after
constructing the hull (post\[hy]merging).
Both pre\[hy] and post\[hy]merging can be defined.


src/Changes.txt
$Id: //main/2019/qhull/src/Changes.txt#87 $$Date: 2024/07/23 $

.............This file lists all changes to qhull and rbox.....................

====
== Contents
====

To do
Qhull 2021-2024
Qhull 2020
Qhull 2018-2019
Qhull 2015
Qhull 2012.1
Qhull 2011.1
Qhull 2010.1 and 2009.1
Qhull 2003.1
Qhull 2002.1
Qhull 3.1 (2001)
Qhull 2.6 (1999)
Qhull 2.5 (1998)
Qhull 2.4 (1997)
Qhull 2.3 (1996)
Qhull 2.2 (1995-1996)
Qhull 2.1 (1995)
Qhull 2.0 (1994-1995)

====
== To do
====

- For a list of enhancement requests, see
  http://www.qhull.org/html/qh-code.htm#enhance
- The C++ interface needs documentation.
  Give C++ a try and make it better.
  http://github.com/qhull/qhull/wiki
- Expand the C++ interface for Voronoi diagrams
  Design for arbitrary dimension with non-simplicial facets
  Start with 'user_eg3 eg-voronoi eg-fifo'
  Compute the convex hull of each Voronoi region
  Compare its facet hyperplanes with eg-fifo
  Create a data structure for the Voronoi diagram
  Compare your result to CGAL's 2-d Voronoi diagram
  Two CGAL halfedges is equivalent to a Qhull ridge
  https://doc.cgal.org/4.7/Voronoi_diagram_2/group__PkgVoronoiDiagramAdaptor2.html

==================
== Qhull 2021-2024
==================

------------
Qhull 8.1.alpha4 GitHub 2024/07/22

Build changes
- Replaced tag 8.1-alpha4 with 8.1.alpha4 [T. Kloczek #106]
- CMakeLists.txt: properly express dependency of qhullcpp to qhullstatic_r  [SpaceIm, M. Valgur #128] 
- qhull-exports.def,*.def: Add DATA after qh_version and qh_version2 [B. Sanders #139]
-  other symbols in *.def are CODE symbols exported to Windows

Documentation changes
- qconvex,qhalf.htm: OFF format 'o' is points instead of vertices [B. Guang #134]
- libqhull_r.h,global_r,merge_r.c: Change Q15-check-duplicates to 'Q17' as coded [djerius #140]
- libqhull.h,global,merge.c: Change Q15-check-duplicates to 'Q17' as coded (pre-Q2019.0.1)
- qconvex_r,qdelaun_r,qhalf_r,unix_r,qvoronoi_r.c::hidden_options: Change Q15 to Q17
- qconvex,qdelaun,qhalf,unix,qvoronoi.c::hidden_options: Change Q15 to Q17
- Qhull::s_not_output_options: Change Q15 to Q17
- q_test-ok.txt,qh-code,qh-optq,qh-quick,qhull.htm,qhull.man,qhull.txt: Change to Q17

------------
Qhull 8.1-alpha3 GitHub 2023/01/02

- user_r.h, user.h: Fix header comment for REALmax,REALmin,REALepsilon
- global_r.c, global.c: Update qh_version2 and new date for qh_version
- CMakeLists.txt: Update qhull_VERSION
- README.txt: CMakeCache.txt is in the build/ directory
- Update Perforce

------------
Qhull 8.1-alpha2 GitHub 2023/01/01

- Reviewed and wrote comments for open issues #121 and earlier.
- Testing postponed to the next release of Qhull

Code changes
- QhullUser.cpp: Redefine snprintf as sprintf_s for VS2010 and earlier [cmF3 #103]
- QhullLinkeList.h, QhullSet.h: Remove unused type parameters from inside a class template [E. Lorenz, J. Nimmer #122, P. Jia] 
- qhull_r-exports.def: Add new symbols from accessors_r.h [J. Heiskala #105]

Build changes
- CMakeLists.txt:
-   Define 'uninstall' target if not previously defined [F. Ramano #111]
-   Add QHULL_ENABLE_TESTING to disable testing if needed [F. Ramano #112]
-   Replace 'install' directory with INCLUDE_INSTALL_DIR [R. Wittmaack #113, O. Poplawski #123]
-   Move 'cmake_minimum_required' to the first instruction [spacelm #120]
-   Use ${LIB_INSTALL_DIR} for the CMake and PkgConfig install directories [O. Poplawski #123]
-   Add BUILD_APPLICATIONS to disable build of Qhull executables [M. Kuhn #124]
- Makefile:
-   Add '-lm' for libqhull_r.so as done elsewhere [G. Zhang #114]
-   Use 'sed' to allow "8.1-alpha1" as qhull_VERSION in CMakeLists.txt [G. Zhang #115]

------------
Qhull 8.1-alpha1 GitHub 2021/09/27
- Tag Qhull #head for GitHub additions to Qhull [S.J. Johnson #97]
- Testing postponed to the next release of Qhull
- global_r.c, global.c: Update qh_version2 and new date for qh_version
- CMakeLists.txt:
-   Update qhull_VERSION and qhull_SOVERSION
-   Change MSVC to WIN32 for dllimport attribute [B. Nath #80]
-   Add BUNDLE DESTINATION to install(QhullTargets) for iOS [spacelm #90]
-   Define CMAKE_INSTALL_NAME_DIR enabling default set_target_properties [B. Boeckel #94]
-   Define CMAKE_INSTALL_RPATH, CMAKE_INSTALL_RPATH_USE_LINK_PATH. CMAKE_BUILD_WITH_INSTALL_RPATH [#94]

Additions to 
- accessors_r.h/.c -- Accessor functions for qhull integration [S. Johnson, B. Legat #89]
- qhull_r-exports.def -- Add some missing symbols: qh_memcheck and qh_setendpointer [F. Facioni #93]

Documentation changes
- user_r.c,user_eg2_r.c: Replace -REALmin with -REALmax in two error messages [S. Hutter #87]
- Fixed various spelling errors in code comments and Changes.txt [luzpaz #84]

==================
== Qhull 2020
==================

------------
Qhull 2020.2 2020/08/31 (8.0.2)

CMake builds
- CMakeLists.txt: Fixed /usr/local/lib/pkgconfig for debug builds
- CMakeLists.txt: Remove qhull_SHARED and qhull_SHAREDP from 'make all'
  Remove qhull_SHARED and qhull_SHAREDP from qhull_TARGETS_SHARED
  Use reentrant Qhull (libqhull_r) for user code
- CMakeLists.txt: If "MSYS Makefiles", change the default CMAKE_INSTALL_PREFIX to "/usr/local"
  Remove WIN32 conditionals from MAN_INSTALL_DIR and DOC_INSTALL_DIR
- CMakeLists.txt: Add 'uninstall' to delete the files listed in install_manifest.txt
- CMakeLists.txt: If LINK_APPS_SHARED, link qhull applications to libqhull_r.so [S. Bruens #69]
  If BUILD_STATIC_LIBS=OFF or BUILD_SHARED_LIBS=OFF, exclude those libraries [S. Bruens #69]
  libqhullcpp and user_eg3 are built only if BUILD_STATIC_LIBS is true
- CMakeModules/CheckLFS.cmake: Moved CMakeModules to the build/ directory
- CMakeLists.txt: Replace rbox_SOURCES(etc.) with a source file, either reentrant or non-reentrant
- CMakeLists.txt: Add deprecated libqhull and qhull_p to "# Additional build targets"
  If not MSVC, add qh_QHpointer to user_egp_DEFINES and qhullp_DEFINES (otherwise qh_lib_check fails)

Other builds
- build/README-build.txt: Add description of build/ directory

src/libqhullcpp/Qhull.cpp
/****************************************************************************
**
** Copyright (c) 2008-2020 C.B. Barber. All rights reserved.
** $Id: //main/2019/qhull/src/libqhullcpp/Qhull.cpp#13 $$Change: 3664 $
** $DateTime: 2024/07/22 23:55:01 $$Author: bbarber $
**
****************************************************************************/

#//! Qhull -- invoke qhull from C++
#//! Compile libqhull_r and Qhull together due to use of setjmp/longjmp()

#include "libqhullcpp/Qhull.h"

#include "libqhullcpp/QhullError.h"
#include "libqhullcpp/RboxPoints.h"
#include "libqhullcpp/QhullQh.h"
#include "libqhullcpp/QhullFacet.h"
#include "libqhullcpp/QhullFacetList.h"

#include <iostream>

using std::cerr;
using std::string;
using std::vector;
using std::ostream;

#ifdef _MSC_VER  // Microsoft Visual C++ -- warning level 4
#pragma warning( disable : 4611)  // interaction between '_setjmp' and C++ object destruction is non-portable
#pragma warning( disable : 4996)  // function was declared deprecated(strcpy, localtime, etc.)
#endif

namespace orgQhull{

#//!\name Global variables

const char s_unsupported_options[]=" Fd TI ";
const char s_not_output_options[]= " Fd TI A C d E H P Qa Qb QbB Qbb Qc Qf Qg Qi Qm QJ Qr QR Qs Qt Qv Qx Qz Q0 Q1 Q2 Q3 Q4 Q5 Q6 Q7 Q8 Q9 Q10 Q11 Q17 R TA Tc TC TM TP TR Tv TV TW U v V W ";

#//!\name Constructor, destructor, etc.
Qhull::
Qhull()
: qh_qh(0)
, origin_point()
, run_called(false)
, feasible_point()
{
    allocateQhullQh();
}//Qhull

//! Invokes Qhull on rboxPoints
//! Same as runQhull()
//! For rbox commands, see http://www.qhull.org/html/rbox.htm or html/rbox.htm
//! For qhull commands, see http://www.qhull.org/html/qhull.htm or html/qhull.htm
Qhull::
Qhull(const RboxPoints &rboxPoints, const char *qhullCommand2)
: qh_qh(0)
, origin_point()
, run_called(false)
, feasible_point()
{
    allocateQhullQh();
    runQhull(rboxPoints, qhullCommand2);
}//Qhull rbox

//! Invokes Qhull on a set of input points
//! Same as runQhull()
//! For qhull commands, see http://www.qhull.org/html/qhull.htm or html/qhull.htm
Qhull::
Qhull(const char *inputComment2, int pointDimension, int pointCount, const realT *pointCoordinates, const char *qhullCommand2)
: qh_qh(0)
, origin_point()
, run_called(false)
, feasible_point()
{
    allocateQhullQh();
    runQhull(inputComment2, pointDimension, pointCount, pointCoordinates, qhullCommand2);
}//Qhull points

void Qhull::
allocateQhullQh()
{
  QHULL_LIB_CHECK /* Check for compatible library */

    qh_qh= new QhullQh;
    void *p= qh_qh;
    void *p2= static_cast<qhT *>(qh_qh);
    char *s= static_cast<char *>(p);
    char *s2= static_cast<char *>(p2);
    if(s!=s2){
        throw QhullError(10074, "Qhull error: QhullQh at a different address than base type QhT (%d bytes).  Please report compiler to qhull.org", int(s2-s));
    }
}//allocateQhullQh

Qhull::
~Qhull() throw()
{
    // Except for cerr, does not throw errors
    if(qh_qh->hasQhullMessage()){
        cerr<< "\nQhull messages at ~Qhull()\n"; // QH11005 FIX: where should error and log messages go on ~Qhull?
        cerr<< qh_qh->qhullMessage();
        qh_qh->clearQhullMessage();
    }
    delete qh_qh;
    qh_qh= 0;
}//~Qhull

#//!\name GetSet

void Qhull::
checkIfQhullInitialized()
{
    if(!initialized()){ // qh_initqhull_buffers() not called
        throw QhullError(10023, "Qhull error: checkIfQhullInitialized failed.  Call runQhull() first.");
    }
}//checkIfQhullInitialized

//! Return feasiblePoint for halfspace intersection
//! If called before runQhull(), then it returns the value from setFeasiblePoint.  qh.feasible_string overrides this value if it is defined.
Coordinates Qhull::
feasiblePoint() const
{
    Coordinates result;
    if(qh_qh->feasible_point){
        result.append(qh_qh->hull_dim, qh_qh->feasible_point);
    }else{
        result= feasible_point;
    }
    return result;
}//feasiblePoint

//! Return origin point for qh.input_dim
QhullPoint Qhull::
inputOrigin()
{
    QhullPoint result= origin();
    result.setDimension(qh_qh->input_dim);
    return result;
}//inputOrigin

#//!\name GetValue

double Qhull::
area(){
    checkIfQhullInitialized();
    if(!qh_qh->hasAreaVolume){
        QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
            qh_getarea(qh_qh, qh_qh->facet_list);
        }
        qh_qh->NOerrexit= true;
        qh_qh->maybeThrowQhullMessage(QH_TRY_status);
    }
    return qh_qh->totarea;
}//area

double Qhull::
volume(){
    checkIfQhullInitialized();
    if(!qh_qh->hasAreaVolume){
        QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
            qh_getarea(qh_qh, qh_qh->facet_list);
        }
        qh_qh->NOerrexit= true;
        qh_qh->maybeThrowQhullMessage(QH_TRY_status);
    }
    return qh_qh->totvol;
}//volume

#//!\name Foreach

//! Define QhullVertex::neighborFacets().
//! Automatically called if merging facets or computing the Voronoi diagram.
//! Noop if called multiple times.
void Qhull::
defineVertexNeighborFacets(){
    checkIfQhullInitialized();
    if(!qh_qh->hasAreaVolume){
        QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
            qh_vertexneighbors(qh_qh);
        }
        qh_qh->NOerrexit= true;
        qh_qh->maybeThrowQhullMessage(QH_TRY_status);
    }
}//defineVertexNeighborFacets

QhullFacetList Qhull::
facetList() const
{
    return QhullFacetList(beginFacet(), endFacet());
}//facetList

QhullPoints Qhull::
points() const
{
    return QhullPoints(qh_qh, qh_qh->hull_dim, qh_qh->num_points*qh_qh->hull_dim, qh_qh->first_point);
}//points

QhullPointSet Qhull::
otherPoints() const
{
    return QhullPointSet(qh_qh, qh_qh->other_points);
}//otherPoints

//! Return vertices of the convex hull in qh.vertex_list order
//! Vertices are not ordered by adjacency (see QhullFacet::nextFacet2d)
QhullVertexList Qhull::
vertexList() const
{
    return QhullVertexList(beginVertex(), endVertex());
}//vertexList

#//!\name Methods

void Qhull::
outputQhull()
{
    checkIfQhullInitialized();
    QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
        qh_produce_output2(qh_qh);
    }
    qh_qh->NOerrexit= true;
    qh_qh->maybeThrowQhullMessage(QH_TRY_status);
}//outputQhull

void Qhull::
outputQhull(const char *outputflags)
{
    checkIfQhullInitialized();
    string cmd(" "); // qh_checkflags skips first word
    cmd += outputflags;
    char *command= const_cast<char*>(cmd.c_str());
    QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
        qh_clear_outputflags(qh_qh);
        char *s= qh_qh->qhull_command + strlen(qh_qh->qhull_command) + 1; //space
        strncat(qh_qh->qhull_command, command, sizeof(qh_qh->qhull_command)-strlen(qh_qh->qhull_command)-1);
        qh_checkflags(qh_qh, command, const_cast<char *>(s_not_output_options));
        qh_initflags(qh_qh, s);
        qh_initqhull_outputflags(qh_qh);
        if(qh_qh->KEEPminArea < REALmax/2 || qh_qh->KEEParea || qh_qh->KEEPmerge || qh_qh->GOODvertex
                  || qh_qh->GOODpoint || qh_qh->GOODthreshold || qh_qh->SPLITthresholds){
            facetT *facet;
            qh_qh->ONLYgood= False;
            FORALLfacet_(qh_qh->facet_list){
                facet->good= True;
            }
            qh_prepare_output(qh_qh);
        }
        qh_produce_output2(qh_qh);
        if(qh_qh->VERIFYoutput && !qh_qh->FORCEoutput && !qh_qh->STOPadd && !qh_qh->STOPcone && !qh_qh->STOPpoint){
            qh_check_points(qh_qh);
        }
    }
    qh_qh->NOerrexit= true;
    qh_qh->maybeThrowQhullMessage(QH_TRY_status);
}//outputQhull

//! Prepare Qhull for Voronoi output
//! Calls qh_markvoronoi ('v o Fi Fo') and qh_order_vertexneighbors ('v Fi Fo')
void Qhull::
prepareVoronoi(bool *isLower, int *voronoiVertexCount)
{
  QH_TRY_(qh_qh){ // no object creation -- destructors skipped on longjmp()
    boolT isLowerHull;
    vertexT *vertex;

    setT *vertices= qh_markvoronoi(qh_qh, facetList().first().getFacetT(), NULL, !qh_ALL, &isLowerHull, voronoiVertexCount);
    *isLower= isLowerHull;

    qhT *qh= qh_qh;
    FORALLvertices{
      qh_order_vertexneighbors(qh, vertex);
    }
    qh_settempfree(qh, &vertices);
  }
  qh_qh->NOerrexit= true;
  qh_qh->maybeThrowQhullMessage(QH_TRY_status);
}//prepareVoronoi

//! For qhull commands, see http://www.qhull.org/html/qhull.htm or html/qhull.htm
void Qhull::
runQhull(const RboxPoints &rboxPoints, const char *qhullCommand2)
{
    runQhull(rboxPoints.comment().c_str(), rboxPoints.dimension(), rboxPoints.count(), &*rboxPoints.coordinates(), qhullCommand2);
}//runQhull, RboxPoints

//! pointCoordinates is a array of points, input sites ('d' or 'v'), or halfspaces with offset last ('H')
//! Derived from qh_new_qhull [user.c]
//! For rbox commands, see http://www.qhull.org/html/rbox.htm or html/rbox.htm
//! For qhull commands, see http://www.qhull.org/html/qhull.htm or html/qhull.htm
void Qhull::
runQhull(const char *inputComment2, int pointDimension, int pointCount, const realT *pointCoordinates, const char *qhullCommand2)
{
  /* gcc may issue a "might be clobbered" warning for pointDimension and pointCoordinates [-Wclobbered].
     These parameters are not referenced after a longjmp() and hence not clobbered.
     See http://stackoverflow.com/questions/7721854/what-sense-do-these-clobbered-variable-warnings-make */
    if(run_called){
        throw QhullError(10027, "Qhull error: runQhull called twice.  Only one call allowed.");
    }
    run_called= true;
    string s("qhull ");
    s += qhullCommand2;
    char *command= const_cast<char*>(s.c_str());
    /************* Expansion of QH_TRY_ for debugging
    int QH_TRY_status;
    if(qh_qh->NOerrexit){
        qh_qh->NOerrexit= False;
        QH_TRY_status= setjmp(qh_qh->errexit);
    }else{
        QH_TRY_status= QH_TRY_ERROR;
    }
    if(!QH_TRY_status){
    *************/
    QH_TRY_(qh_qh){ // no object creation -- destructors are skipped on longjmp()
        qh_checkflags(qh_qh, command, const_cast<char *>(s_unsupported_options));
        qh_initflags(qh_qh, command);
        *qh_qh->rbox_command= '\0';
        strncat( qh_qh->rbox_command, inputComment2, sizeof(qh_qh->rbox_command)-1);
        if(qh_qh->DELAUNAY){
            qh_qh->PROJECTdelaunay= True;   // qh_init_B() calls qh_projectinput()
        }
        pointT *newPoints= const_cast<pointT*>(pointCoordinates);
        int newDimension= pointDimension;
        int newIsMalloc= False;
        if(qh_qh->HALFspace){
            --newDimension;
            initializeFeasiblePoint(newDimension);
            newPoints= qh_sethalfspace_all(qh_qh, pointDimension, pointCount, newPoints, qh_qh->feasible_point);
            newIsMalloc= True;
        }
        qh_init_B(qh_qh, newPoints, pointCount, newDimension, newIsMalloc);
        qh_qhull(qh_qh);
        qh_check_output(qh_qh);
        qh_prepare_output(qh_qh);
        if(qh_qh->VERIFYoutput && !qh_qh->FORCEoutput && !qh_qh->STOPadd && !qh_qh->STOPcone && !qh_qh->STOPpoint){
            qh_check_points(qh_qh);
        }
    }
    qh_qh->NOerrexit= true;
    for(int k= qh_qh->hull_dim; k--; ){  // Do not move into QH_TRY block.  It may throw an error
        origin_point << 0.0;
    }
    qh_qh->maybeThrowQhullMessage(QH_TRY_status);
}//runQhull

#//!\name Helpers -- be careful of allocating C++ objects due to setjmp/longjmp() error handling by qh_... routines

//! initialize qh.feasible_point for half-space intersection
//! Sets from qh.feasible_string if available, otherwise from Qhull::feasible_point
//! called only once from runQhull(), otherwise it leaks memory (the same as qh_setFeasible)
void Qhull::
initializeFeasiblePoint(int hulldim)
{
    if(qh_qh->feasible_string){
        qh_setfeasible(qh_qh, hulldim);
    }else{
        if(feasible_point.isEmpty()){
            qh_fprintf(qh_qh, qh_qh->ferr, 6209, "qhull error: missing feasible point for halfspace intersection.  Use option 'Hn,n' or Qhull::setFeasiblePoint before runQhull()\n");
            qh_errexit(qh_qh, qh_ERRmem, NULL, NULL);
        }
        if(feasible_point.size()!=static_cast<size_t>(hulldim)){
            qh_fprintf(qh_qh, qh_qh->ferr, 6210, "qhull error: dimension of feasiblePoint should be %d.  It is %u\n", hulldim, (unsigned int)feasible_point.size());
            qh_errexit(qh_qh, qh_ERRmem, NULL, NULL);
        }
        qh_qh->feasible_point= static_cast<coordT*>(qh_malloc(static_cast<size_t>(hulldim) * sizeof(coordT)));
        if(!qh_qh->feasible_point){
            qh_fprintf(qh_qh, qh_qh->ferr, 6042, "qhull error (Qhull.cpp): insufficient memory for feasible point\n");
            qh_errexit(qh_qh, qh_ERRmem, NULL, NULL);
        }
        coordT *t= qh_qh->feasible_point;
        // No qh_... routines after here -- longjmp() ignores destructor
        for(Coordinates::ConstIterator p=feasible_point.begin(); p<feasible_point.end(); p++){
            *t++= *p;
        }
    }
}//initializeFeasiblePoint

}//namespace orgQhull

